src/api/OffsetBasedAPI.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * @flow
+ * @file class for Box offset based API's to inherit common functionality from
+ * @author Box
+ */
+import Base from './Base';
+
+type Params = {
+    offset: number,
+    limit: number,
+    fields?: string
+};
+
+type Data = {
+    entries: Array<any>,
+    total_count: number
+};
+
+class OffsetBasedApi extends Base {
+    /**
+     * @property {Data}
+     */
+    data: Data;
+
+    /**
+     * Gets query params for the API
+     *
+     * @return the query params object
+     */
+    getQueryParameters(offset: number, limit: number, fields: ?Array<string>): Object {
+        const queryParams: Params = {
+            offset,
+            limit
+        };
+
+        if (fields && fields.length > 0) {
+            queryParams.fields = fields.toString();
+        }
+
+        return queryParams;
+    }
+
+    /**
+     * Determines if the API has more items to fetch
+     * @param {number} offset the offset from the start to start fetching at
+     * @param {number} totalCount the total number of items
+     * @return {boolean} true if there are more items
+     */
+    hasMoreItems(offset: number, totalCount?: number): boolean {
+        return totalCount === undefined || offset < totalCount;
+    }
+
+    /**
+     * Offset based API get
+     *
+     * @override
+     * @param {string} id the file id
+     * @param {Function} successCallback the success callback
+     * @param {Function} errorCallback the error callback
+     * @param {number} offset the offset from the start to start fetching at
+     * @param {number} limit the number of items to fetch
+     * @param {array} fields the fields to fetch
+     * @param {boolean} shouldFetchAll true if should get all the pages before calling the sucessCallback
+     */
+    async get(
+        id: string,
+        successCallback: Function,
+        errorCallback: Function,
+        offset: number = 0,
+        limit: number = 1000,
+        fields?: Array<string>,
+        shouldFetchAll?: boolean
+    ): Promise<void> {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.successCallback = successCallback;
+        this.errorCallback = errorCallback;
+
+        // Make the XHR request
+        try {
+            const params = this.getQueryParameters(offset, limit, fields);
+            const newOffset = offset + limit;
+
+            const { data }: { data: Data } = await this.getData(id, params);
+
+            const entries = this.data ? this.data.entries : [];
+            this.data = {
+                ...data,
+                entries: entries.concat(data.entries)
+            };
+            const totalCount = data.total_count;
+
+            if (shouldFetchAll && this.hasMoreItems(newOffset, totalCount)) {
+                this.get(id, successCallback, errorCallback, newOffset, limit, fields, shouldFetchAll);
```

Comment:
need to pass in success/error callback if its already on the `this` object?

Reply:
In this case, yeah since its a recursive call and it sets the this variables in the function. One way to get rid of the args I could create a helper method which takes all the args, and create a helper function without the callbacks

Reply:
Refactored this method a bit, ended up renaming it to offsetGet and using a helper method